### DISCUSSION TITLE
WRITE_BUFFER_SIZE set as 1024 where it was 2048(DATA_BUFFER_SIZE) in memcached 1.5.22

### DISCUSSION THREAD

**Discussion by ngoyal1987** [memcached/memcached#1179] [Q&A] [answered]
Can someone please explain the reason to reduce the Data Buffer size from 2048 to 1024 in the versions after 1.5.22. It used to be DATA_BUFFER_SIZE 2048 in version 1.5.22 and earlier. I am currently working on patching repcached over the latest memcached version 1.6.31 (upgrade from 1.5.22).

**Comment by dormando** [comment, marked as answer]
The whole network system changed, not just the size of that buffer. There used to be a static write buffer on each client connection, which started at DATA_BUFFER_SIZE but could dynamically grow. That's been replaced by a linked list of temporary smaller buffers (fixed WRITE_BUFFER_SIZE).

This was done to speed up writes and reduce memory usage of idle tcp connections, since we no longer hold onto buffers when a conn is idle.

You might not need repcached anymore, if you can use the proxy instead: https://docs.memcached.org/features/proxy/ - it can copy set commands to multiple backend pools, which will both scale better and is more flexible than the old repcached. It does not synchronize data between hosts however.

**Reply by ngoyal1987** [reply]
Thanks a lot @dormando for your answer. This means now we need not to care for "replication_alloc"(used to increase the wbuf size in case the data is large) since the write buffer will accommodate dynamically. 
Also how we now find the pointer to the current location in this new write buffer(wbuf in mc_resp).

Thanks for suggesting the proxy solution. I will analyze this solution to see if it is best fit for us. But as you mentioned that it doesn't synchronize the data between hosts. Then what can we use to achieve high availability in case of host down.

**Reply by dormando** [reply]
The proxy can write the same data to multiple pools or hosts. It does not do an initial data load. If you are using a cache data pattern this is fine: The cache will fill over time and the proxy can "fail over" between pools looking for a key.

To explain another way: If you start from scratch with two memcached hosts, and use the proxy to access them, it can copy all writes to both hosts, and will achieve high availability on reads. In this case both hosts will be in sync.

If later one host breaks and you replace it, it will come up empty. Over time the host will re-fill as your data data expires and gets re-set to the hosts. The proxy will work with this completely fine.

If the data needs to be 100% synchronized please explain your situation in more detail. There are probably simpler ways of implementing something like repcached now. The old patch is very very old and memcached has changed a lot since then.

**Reply by ngoyal1987** [reply]
Our application is a low latency application which has 2 hosts architecture operating as Active-HotStandby with stringent failover SLA. Memcached is getting replicated on both the hosts. If one hosts breaks, then within seconds, the standby hosts takes over to become active and starts serving. So it is important for us to have both the hosts have data in sync with each other all the time.

Yes, old Repcached patch is old. For our implementation for 1.6.31, we are now left with finding the pointer to the current location in this new write buffer(wbuf in mc_resp) like it was pointed by "wcurr" in old implementation.

**Reply by dormando** [reply]
I'm not familiar with the repcached patch, that doesn't sound like you're on the right path though.

You should take a stronger consideration to using the proxy to handle set replication and failover. It will fail over instantly, rather than within a few seconds. It will internally retry requests that failed due to a down server.

We have data dumping and loading facilities by default now, so you don't need repcached for this either.

In a scenario where you have servers A and B:
- If A/B are started at the same time, empty, and you use the proxy to keep them in sync everything is happy.
- Server A dies. The proxy now serves requests from B.
- Server A is replaced, and comes up empty.
- The proxy tries to fetch data from A, but it is not there, and it will fall back to B.

In this scenario, you can use a tool like: https://github.com/memcached/mcdumpload as part of your maintenance scripts. Once A comes up, run mcdumpload from B with a target of A. Once it is complete your data should be close to in sync. The proxy handles replicating all new SET requests while this process runs.

I have some other dumpload utilities but they're less user friendly. mcdumpload is extremely fast and efficient. I can write more documentation for it if needed.

edit: this is also a more scalable solution. If you want to go from two replicated servers, to two replicated _pools_, of say three memcached's each, the proxy will let you do this. The mcdumpload part can become more complicated in this scenario but it's possible.

**Reply by dormando** [reply]
You didn't say how your data is loaded in the first place or if it's cache data or fixed data btw? If it's a cache pattern you can still be okay with just the proxy and no dumpload, since it will try the other server if a GET results in a miss. Eventually caches get refreshed and things will be in sync again.

It will also soon be possible for the proxy to recache on miss. For example:

- Servers A/B.
- A becomes empty.
- Proxy fetches from A, gets a miss.
- Proxy fetches from B, returns result to client. Asynchronously sets the data to A.
- Active data slowly resyncs.

**Reply by ngoyal1987** [reply]
Hello, sorry for the delayed response. I am not sure how exactly to categorize the data. It's runtime data which is fixed for request processing, but keeps changing per requests. But not needed to be stored once processing is done.

**Reply by dormando** [reply]
Ah sorry I didn't see your response. It sounds like your data is relatively short lived? Each piece of data is only alive for a few seconds?

You probably don't need to re-synchronize the servers at all then. If you fail over and replace the dead node the new one will be in sync after a few more seconds, and hopefully the other one won't die in the meantime :) With the proxy you could even have three of them.

**Reply by tommey** [reply]
As I come by this discussion due to the same task "update repcached": we started from 1.4.13, I managed to merge all tags until 1.5.22, had to keep in the inline_ascii_reply feature and wrap all c->thread usages, it passed smoke test, then went to try 1.6.0, merged it, of course it died due to changed usages in replication.c, while I was gathering energy to continue I went on an read all release notes, then I found 1.6.36 stating there is an **internal caching feature in the proxy**, which is currently not documented, went back straight to the code and found it. Earlier I was also messing with the possibility of implementing it myself with some highly hackis socket virtualization workaround, then I also stumbled upon the undocumented feature that **protocol can be forced on listen** addresses.

So now thanks to these, it is possible to configure memcached to act like repcached with:
* memcached u memcache -d -m 2048 -l proto[proxy]:0.0.0.0:11211,proto[negotiating]:0.0.0.0:11212 -o proxy_config=routelib,proxy_arg=routing.lua

With routing.lua:
* pools{pair={backends={"<pair IP>:11212"}}} routes{cmap={get=route_direct{child="internal"},gets=route_direct{child="internal"}},default=route_split{child_a="internal",child_b="pair"}}

My issue with the standalone proxy instance was that it requires +100% cpu just to forward each command to the local caching instance, also we would need to configure and monitor two instances per server; with the above solution it only needs extra CPU for the commands that are forwarded to the pair, in a 90% read 10% write scenario it is around 20% on a 2 vcpu 8gb ram Azure machine.

We have very small caches, used for http call caching and token-kind of short lived values, while a failure isn't a big deal business-wise, in some cases it would cause a lot of extra http calls and end-users failing actions once, so repcached simply solved these by being able to failover to the pair and have around up-to-date data decreasing the misses; we also use both instances with key sharding to split the load.

Otherwise we were considering switching to Redis, but it requires much more infrastructure with sentinel clusters and only provides one write instance; it would surely work, but this simple memcached replication is just good enough and much simpler.

Also great work on the slab reallocation changes, that's what triggered the whole topic, because 1.4.13 was the last version that didn't do anything about it and after years now we have instances with 6MB content and 2GB allocated, evicting new items continuously because the value sizes grew to slab 4 size, but slab 2 is hogging all memory.

Thank you @dormando for all these great new features!
